### PR TITLE
Fix 'invalid magic number 0' bug

### PR DESCRIPTION
**Pull request by codesome**
Closes https://github.com/prometheus/prometheus/issues/7397

We already had a repair mechanism for this file if the file was empty, but it can happen that file is not 0 size but filled with 0s. So this PR extends the repair logic and repairs for file with 0s as well.

I would like this to be a part of v2.39

### CODE REVIEW

**Comment by alvinlin123** [comment]
I think this fix is good to fix 0 magic number. But I also think we should maybe call `file.Sync()` before closing the `.tmp` file around [when cutting a new segment file](https://github.com/prometheus/prometheus/blob/448cfda6c1e8a98794eb24ae99b7213355111096/tsdb/chunks/chunks.go#L333). 

From what I've read, cutting a new segment file in head chunk involves 
1. create a `.tmp` file
1. pre-allocate the file to some size and fill the file with `0x00`s
1. write magic number 
1. close file
1. rename file

According to Linux's [close man page](https://man7.org/linux/man-pages/man2/close.2.html) data are not guaranteed to be written to disk, and I think Go's [File.Close](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/os/file_unix.go;l=266;drc=361f5eba9f9e9902226e7edac76646253b7025e4) pretty much just make a underlying OS system call. 

I don't think the `fileutil.Rename()` sync the data to disk too. 

I think in this PR maybe we can add a `File.Sync()` call after writing the magic number, then we can do some log grepping on to see if ` invalid magic number 0`  error still happens?

File: tsdb/chunks/head_chunks.go [hunk 1]
```diff
@@ -376,7 +376,29 @@ func repairLastChunkFile(files map[int]string) (_ map[int]string, returnErr erro
...
+			return files, errors.Wrap(err, "open file during last head chunk file repair")
+		}
+
+		buf := make([]byte, MagicChunksSize)
+		_, err = f.Read(buf)
+		if err != nil {
+			return files, errors.Wrap(err, "read file during last head chunk file repair")
+		}
+
+		remove = binary.BigEndian.Uint32(buf) == 0
```
Comment (alvinlin123) [inline tsdb/chunks/head_chunks.go:394]: I think we should log something here to tell the user that "hey, we encountered magic number is 0 error".

**Comment by alvinlin123** [comment]
Ok I see the argument for not doing a sync after writing magic number is noted in https://github.com/prometheus/prometheus/pull/8061#issuecomment-708967115

However, I wonder sync 4 bytes of magic number would take seconds; if it indeed take seconds, isn't the disk overloaded and you have bigger problem?

File: tsdb/chunks/head_chunks.go [hunk 2]
```diff
@@ -376,7 +376,29 @@ func repairLastChunkFile(files map[int]string) (_ map[int]string, returnErr erro
...
+		// Check if magic number is 0.
+		f, err := os.Open(files[lastFile])
+		if err != nil {
+			return files, errors.Wrap(err, "open file during last head chunk file repair")
+		}
+
+		buf := make([]byte, MagicChunksSize)
+		_, err = f.Read(buf)
+		if err != nil {
+			return files, errors.Wrap(err, "read file during last head chunk file repair")
```
Comment (alvinlin123) [inline tsdb/chunks/head_chunks.go:391]: I think it would be nice if the error is more specific: `failed to read magic number during last head chunk file repair`. The target audience of the log message is probably operators, so it's ok to mention magic number.

File: tsdb/chunks/head_chunks.go [hunk 3]
```diff
@@ -376,7 +376,29 @@ func repairLastChunkFile(files map[int]string) (_ map[int]string, returnErr erro
 	if err != nil {
 		return files, errors.Wrap(err, "file stat during last head chunk file repair")
 	}
-	if info.Size() == 0 {
+
+	remove := info.Size() == 0 // Empty file.
+	if info.Size() >= MagicChunksSize {
```
Comment (alvinlin123) [inline tsdb/chunks/head_chunks.go:381]: nit: I think it might be better to move this if block to a helper function so the code can read like: 

```
remove := info.Size() ==0 
remove := isFileCorrupted(...)

if remove {
 ....
}
```

**Comment by codesome** [comment]
It is not just about syncing the magic number. In some OS the whole file is preallocated even before you write the magic number (for example Windows). So any abrupt shutdown before you sync the magic number still needs to be detected (you can see that we already had a code in place to detect empty files, but that is apparently not enough).

Also, fsync being on the hot path can be a dealbreaker (though with the addition of async queue and another open work, it won't be in hot path in future). This repair mechanism is acceptable and I don't see it as a maintainence burden here; a repair mechanism for database is desirable instead of a hard fail IMO (you cannot expect your hardware to work properly all the time, issues happen).

Additionally, if there was no magic number, it also means there were no chunks flushed to the file, so flushing the magic number does not help us in any way. If you see, trying to sync magic number still leaves a room for the error to occur.

File: tsdb/chunks/head_chunks.go (hunk 3 above)
Comment (codesome) [reply tsdb/chunks/head_chunks.go:381]: Since this logic is only used here and that `repairLastChunkFile` is not too huge, I prefer to not create another function here.

File: tsdb/chunks/head_chunks.go (hunk 2 above)
Comment (codesome) [reply tsdb/chunks/head_chunks.go:391]: Thanks for the suggestion, makes sense

File: tsdb/chunks/head_chunks.go (hunk 1 above)
Comment (codesome) [reply tsdb/chunks/head_chunks.go:394]: Hmm, I don't think adding this log line is useful or actionable. Also, this package does not log anywhere and I would like to keep it that way. Deleting the last file is not a big issue.

Comment (alvinlin123) [reply tsdb/chunks/head_chunks.go:394]: Yea, if we don't want to do fsync I don't think the log will be useful too.

File: tsdb/chunks/head_chunks.go (hunk 3 above)
Comment (alvinlin123) [reply tsdb/chunks/head_chunks.go:381]: Make sense.

**Comment by alvinlin123** [comment]
@codesome yes, you are absolutely right, what you did in this PR is needed; I wouldn't remove it, it's awesome fix :) 

> In some OS the whole file is preallocated even before you write the magic number (for example Windows). So any abrupt shutdown before you sync the magic number still needs to be detected

Right, when you preallocate, I think different OS may do different thing, some may fill it with `0x00`, some may leave the file empty, but that's not the problem here.  The current code is we pre-allocate a temporary file, write the magic number, then rename the file. If we do a fsync before renaming the file, the rename the file, then anything abrupt termination before renaming the file will result the file remain temporary, and will get cleaned up/ignored. In other words, if we can guarantee that before renaming the file, the magic number is written to the temporary file on disk, the the abrupt termination should not result in corrupted segment file (the temporary can still be corrupted, but it doesn't matter).

If we don't do fysnc, then I can see how the actual segment file can be empty, filled with `0x00`, or maybe filled with some dummy value like `0xFF` depending on the underlying OS' implementation of file preallocation.

In other words, I think if we do fsync before renaming the file, we should avoid empty file problem as well as magic number being 0 problem. 
 
> Also, fsync being on the hot path can be a dealbreaker (though with the addition of async queue and another open work, it won't be in hot path in future

My question is more on why having fsync on hot path is a dealbreaker, is the fsync performance really that bad for typical use cases? If we can do some quick perf test on it it would be nice. To put it more bluntly, I wonder if not doing a fsync is premature optimization, but of course I know I can be wrong :-)  The argument of fsync can hurt performance totally make sense, but I am just if it really hurts "that much" :)

**Comment by codesome** [comment]
It is already a problem without fsync on high loads: when the ChunkDiskMapper is trying to flush/mmap many chunks to the disk and it sometimes causes high write latencies because many chunks are waiting to be flushed (this happens on hot path). Trying to fsync (which can happen when writing chunks because you need to go to the next file) will only make it worse.

Again, file getting corrupted like this during abrupt shutdown is not a problem, we can gracefully recover from it without any data loss. Also, trying to fsync does not take the problem away: the abrupt shutdown can happen right when trying to fsync. So I don't see an benefit of a fsync here.

File: tsdb/chunks/head_chunks.go [hunk 4]
```diff
@@ -376,7 +376,29 @@ func repairLastChunkFile(files map[int]string) (_ map[int]string, returnErr erro
...
+			return files, errors.Wrap(err, "open file during last head chunk file repair")
+		}
+
+		buf := make([]byte, MagicChunksSize)
+		_, err = f.Read(buf)
+		if err != nil {
+			return files, errors.Wrap(err, "failed to read magic number during last head chunk file repair")
+		}
+
+		remove = binary.BigEndian.Uint32(buf) == 0
```
Comment (csmarchbanks) [inline tsdb/chunks/head_chunks.go:394]: Should this remove anything that is not `MagicHeadChunks`? I would imagine an incorrect magic number would also be a sign of corruption, not just 0.

File: tsdb/chunks/head_chunks.go [hunk 5]
```diff
@@ -376,7 +376,29 @@ func repairLastChunkFile(files map[int]string) (_ map[int]string, returnErr erro
 	if err != nil {
 		return files, errors.Wrap(err, "file stat during last head chunk file repair")
 	}
-	if info.Size() == 0 {
+
+	remove := info.Size() == 0 // Empty file.
+	if info.Size() >= MagicChunksSize {
+		// Check if magic number is 0.
+		f, err := os.Open(files[lastFile])
```
Comment (csmarchbanks) [inline tsdb/chunks/head_chunks.go:383]: Do we still need the os.Stat call above if we are almost always going to open the file? OK with it as is though.

File: tsdb/chunks/head_chunks.go (hunk 4 above)
Comment (codesome) [reply tsdb/chunks/head_chunks.go:394]: If the magic number was fully present and not equals then it is caught at some other place that checks for file corruptions, so it should be already taken care of.

Comment (codesome) [reply tsdb/chunks/head_chunks.go:394]: Looks like the snapshot replay relies on the genuine corruption being reported up the function call so that we can discard the snapshot. So I think we should not take care of different magic number here. However, this gave me an idea that we are missing another case: if the number of bytes in the file was 1-3, then we say it is not corrupted while it should be deleted here (because size is not 0 and we dont check magic number in that case). I will fix that.

File: tsdb/chunks/head_chunks.go (hunk 5 above)
Comment (codesome) [reply tsdb/chunks/head_chunks.go:383]: I was not sure how to check for the length with file open, turns out f.Read will return io.EOF with bytes read which should help here.

**Review by csmarchbanks — APPROVED** [review]

File: tsdb/chunks/head_chunks.go (hunk 4 above)
Comment (csmarchbanks) [reply tsdb/chunks/head_chunks.go]: Good to know about the genuine corruption being handled elsewhere, thanks for adding the comment.

**Comment by alvinlin123** [comment]
> It is already a problem without fsync on high loads: when the ChunkDiskMapper is trying to flush/mmap many chunks to the disk and it sometimes causes high write latencies because many chunks are waiting to be flushed (this happens on hot path). Trying to fsync (which can happen when writing chunks because you need to go to the next file) will only make it worse.

I see, thanks for the follow up :)
